keps/sig-api-machinery/2608-crd-field-level-versioning/README.md

```diff
@@ -0,0 +1,816 @@
+- If, now a non-gated field `f'` is updated, the controller gets an event (assuming it is using the usual watch and informer mechanism).
+- The controller sees that `f` has value `v'`, it tries to reconcile it to value `v`, but the update has no effect as `g` is disabled.
+
+To mitigate this scenario, we recommend making use of the `generation` field in the object's `metadata` to check if an update was actually made or not. Please see the Implementation section for more details on this.
+
+Subsequently, if you are using a client such as `kubectl` in the above scenario, if an update does not actually happen, an additional server side warning goes out conveying this and can be displayed back to the user.
+
+## Design Details
+
+### API Changes
+The API changes include extending the CRD spec by adding an optional field:
+```go
+// CustomResourceDefinitionSpec describes how a user wants their resource to appear
+type CustomResourceDefinitionSpec struct {
+    ...
+    // customFeatureGates defines feature gates for fields of
+    // custom resources of this CRD.
+    // +optional
+    CustomFeatureGates *CustomResourceDefinitionFeatureGates
+}
+```
+Additionally, a few types are introduced:
+```go
+// CustomResourceDefinitionFeatureGates defines feature gates for specific
+// fields to enable field level versioning for custom resource definitions.
+type CustomResourceDefinitionFeatureGates struct {
+    // featureGates is a list of gates defined for fields of a custom
+    // resource definition. No two feature gates may guard the same field.
+    FeatureGates []CustomResourceDefinitionFeatureGate
+    // component is meant to indicate the entity that is responsible for
+    // the creation of these feature gates when the feature gate information
+    // is published to OpenAPI.
+    // +optional
+    Component *string
+}
+```
+```go
+// CustomResourceDefinitionFeatureGate describes the information
+// conveyed by a feature gate.
+type CustomResourceDefinitionFeatureGate struct {
+    // name is the name of the feature gate being defined.
+    Name string
+    // enabled signifies whether the feature gatea is enabled or not.
+    // +optional
+    Enabled *bool
+    // default is the default enablement state for the feature.
+    // +optional
+    Default *bool
+    // preRelease indicates the stage of development that the
+    // field/fields it is guarding is in. Acceptable values are
+    // alpha, beta, stable, deprecated.
+    PreRelease string
+    // fieldDeprecationWarning overrides the default warning returned
+    // to API clients when a field marked as deprecated is used. This
+    // may only be set when `deprecated` is true. The default warning
+    // indicates the field path of this field along with messaging that
+    // it is deprecated.
+    // +optional
+    FieldDeprecationWarning *string
+    // fieldPaths defines the fields in JSON path format that this
+    // feature gate helps gate.
+    FieldPaths []string
```

Comment:
This needs a formal definition.

As commented above, having this as a gate reference inside the OpenAPI spec would be much simpler.